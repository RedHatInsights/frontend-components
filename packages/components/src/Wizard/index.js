export { default } from './Wizard';
export { default as Wizard } from './Wizard';
