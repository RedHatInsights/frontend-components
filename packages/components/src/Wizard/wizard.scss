.pf-c-modal-box.ins-c-wizard {
    min-height: 75vh;
    min-width: 75vw;
    
    // Style override, this max-height is actually set above by the overall wizard height
    .pf-c-modal-box__body { max-height: none; }
}