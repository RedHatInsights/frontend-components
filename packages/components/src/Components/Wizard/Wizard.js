import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { Modal, Button } from '@patternfly/react-core';

import './wizard.scss';

class Wizard extends Component {
    state = {
        currentStep: 0
    }

    handleNextModalStep = () => {
        this.setState(({ currentStep }) => ({
            currentStep: currentStep + 1
        }));
    }

    handlePreviousModalStep = () => {
        this.setState(({ currentStep }) => ({
            currentStep: currentStep - 1
        }));
    }

    // On modal close, reset currentStep back to the initial step, the call modalToggle(PF)
    handleOnClose = (submit) => {
        const result = this.props.onClose(submit);

        if (result === false) {
            return;
        }

        this.setState({ currentStep: 0 });
    }

    handleKeyPress = (event) => {
        if (event.key === 'Enter' && this.props.isValidated === true) {
            this.handleNextModalStep();
        }
    }

    render() {

        const { isLarge, title, className, isOpen, isValidated, confirmAction, ...props } = this.props;

        const wizardClasses = classNames(
            'ins-c-wizard',
            className
        );

        const renderModalActions =  [
            // Conditionally render 'confirm' button if on last page
            this.state.currentStep < this.props.content.length - 1
                ? <Button
                    key="Next"
                    action="next"
                    variant="primary"
                    isDisabled={ !isValidated }
                    onClick={ this.handleNextModalStep }>
                        Next
                </Button>
                : <Button
                    key="confirm"
                    action="confirm"
                    variant="primary"
                    isDisabled={ !isValidated }
                    onClick={ () => this.handleOnClose(true) }>
                    { confirmAction }
                </Button>,
            // Conditionally render 'previous' button if not on first page
            this.state.currentStep !== 0 &&
                <Button key="back" action="back" variant="secondary" onClick={ this.handlePreviousModalStep }> Back </Button>,
            <Button key="cancel" action="cancel" variant="secondary" onClick={ () => this.handleOnClose(false) }>
                Cancel
            </Button>
        ];

        return (
            <Modal
                { ...props }
                onKeyPress={ this.handleKeyPress }
                isLarge = { isLarge }
                title= { title }
                className= { wizardClasses }
                isOpen={ isOpen }
                onClose={ () => this.handleOnClose(false) }
                actions={ renderModalActions }>
                { this.props.content[this.state.currentStep] }
            </Modal>
        );
    }
}

Wizard.propTypes = {
    isValidated: PropTypes.bool,
    isLarge: PropTypes.bool,
    title: PropTypes.string,
    className: PropTypes.string,
    confirmAction: PropTypes.string,
    isOpen: PropTypes.any,
    content: PropTypes.array,
    onClose: PropTypes.func
};

Wizard.defaultProps = {
    onClose: f => f,
    isValidated: true,
    confirmAction: 'Confirm'
};

export default Wizard;
