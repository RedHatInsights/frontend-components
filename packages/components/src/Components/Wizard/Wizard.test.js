import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Wizard from './Wizard';

const content = [
    <div id="first" key="first">First</div>,
    <div id="second" key="second">Second</div>,
    <div id="third" key="third">Third</div>,
    <div id="fourth" key="fourth">Fourth</div>
];

describe('Wizard component', () => {
    describe('should render correctly', () => {
        it('default', () => {
            const wrapper = shallow(<Wizard content={content} title="Some title" isOpen />);
            expect(toJson(wrapper)).toMatchSnapshot();
        });

        it('large', () => {
            const wrapper = shallow(<Wizard content={content} isLarge title="Some title" isOpen />);
            expect(toJson(wrapper)).toMatchSnapshot();
        });

        it('one page', () => {
            const wrapper = shallow(<Wizard content={[ <div key="one">One</div> ]} title="Some title" isOpen />);
            expect(toJson(wrapper)).toMatchSnapshot();
        });

        it('one page confirm button', () => {
            const wrapper = mount(<Wizard content={[ <div key="one">One</div> ]} title="Some title" isOpen />);
            expect(wrapper.find('[action="confirm"]').first().text()).toBe('Confirm');
        });

        it('one page save button', () => {
            const wrapper = mount(<Wizard content={[ <div key="one">One</div> ]} confirmAction="Save" title="Some title" isOpen />);
            expect(wrapper.find('[action="confirm"]').first().text()).toBe('Save');
        });
    });

    describe('content change', () => {
        it('should happen after clicking on next', () => {
            const wrapper = mount(<Wizard content={content} title="Some title" isOpen />);
            wrapper.find('[action="next"]').first().simulate('click');
            wrapper.update();
            expect(wrapper.find('#second').length).toBe(1);
            expect(wrapper.find('#first').length).toBe(0);
        });

        it('should happen at last content', () => {
            const wrapper = mount(<Wizard content={content} title="Some title" isOpen />);
            wrapper.find('[action="next"]').first().simulate('click');
            wrapper.update();
            wrapper.find('[action="next"]').first().simulate('click');
            wrapper.update();
            wrapper.find('[action="next"]').first().simulate('click');
            wrapper.update();
            expect(wrapper.find('#fourth').length).toBe(1);
            expect(wrapper.find('[action="next"]').length).toBe(0);
            expect(wrapper.find('.pf-c-modal-box__footer > [action="confirm"]').length).toBe(1);
        });

        it('should happen at last content', () => {
            const wrapper = mount(<Wizard content={content} title="Some title" isOpen />);
            wrapper.find('[action="next"]').first().simulate('click');
            wrapper.update();
            wrapper.find('[action="next"]').first().simulate('click');
            wrapper.update();
            expect(wrapper.find('#third').length).toBe(1);
            wrapper.find('[action="back"]').first().simulate('click');
            expect(wrapper.find('#second').length).toBe(1);
        });
    });

    describe('API', () => {
        it('should not call onClose X button', () => {
            const onClose = jest.fn();
            const wrapper = mount(<Wizard content={content} title="Some title" isOpen />);
            wrapper.find('[aria-label="Close"]').first().simulate('click');
            expect(onClose.mock.calls.length).toBe(0);
        });

        it('should not call onClose on close button', () => {
            const onClose = jest.fn();
            const wrapper = mount(<Wizard content={content} title="Some title" isOpen />);
            wrapper.find('[action="cancel"]').first().simulate('click');
            expect(onClose.mock.calls.length).toBe(0);
        });

        it('should call onClose X button', () => {
            const onClose = jest.fn();
            const wrapper = mount(<Wizard content={content} title="Some title" isOpen onClose={onClose}/>);
            wrapper.find('[aria-label="Close"]').first().simulate('click');
            expect(onClose.mock.calls.length).toBe(1);
        });

        it('should call onClose X button', () => {
            const onClose = jest.fn();
            const wrapper = mount(<Wizard content={content} title="Some title" isOpen onClose={onClose} />);
            wrapper.find('[action="cancel"]').first().simulate('click');
            expect(onClose.mock.calls.length).toBe(1);
        });

        it('should call close on confirm', () => {
            const onClose = jest.fn();
            const wrapper = mount(<Wizard content={content} title="Some title" isOpen onClose={onClose} />);
            wrapper.find('[action="next"]').first().simulate('click');
            wrapper.update();
            wrapper.find('[action="next"]').first().simulate('click');
            wrapper.update();
            wrapper.find('[action="next"]').first().simulate('click');
            wrapper.update();
            wrapper.find('[action="confirm"]').first().simulate('click');
            expect(onClose.mock.calls.length).toBe(1);
        });
    });
});
