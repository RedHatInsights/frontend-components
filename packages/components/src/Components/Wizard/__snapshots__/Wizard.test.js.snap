// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wizard component should render correctly default 1`] = `
<Modal
  actions={
    Array [
      <Unknown
        action="cancel"
        onClick={[Function]}
        variant="secondary"
      >
        Cancel
      </Unknown>,
      false,
      <Unknown
        action="next"
        isDisabled={false}
        onClick={[Function]}
        variant="primary"
      >
        Next
      </Unknown>,
    ]
  }
  appendTo={<body />}
  ariaDescribedById=""
  className="ins-c-wizard"
  content={
    Array [
      <div
        id="first"
      >
        First
      </div>,
      <div
        id="second"
      >
        Second
      </div>,
      <div
        id="third"
      >
        Third
      </div>,
      <div
        id="fourth"
      >
        Fourth
      </div>,
    ]
  }
  hideTitle={false}
  isFooterLeftAligned={false}
  isLarge={false}
  isOpen={true}
  isSmall={false}
  onClose={[Function]}
  onKeyPress={[Function]}
  showClose={true}
  title="Some title"
>
  <div
    id="first"
  >
    First
  </div>
</Modal>
`;

exports[`Wizard component should render correctly large 1`] = `
<Modal
  actions={
    Array [
      <Unknown
        action="cancel"
        onClick={[Function]}
        variant="secondary"
      >
        Cancel
      </Unknown>,
      false,
      <Unknown
        action="next"
        isDisabled={false}
        onClick={[Function]}
        variant="primary"
      >
        Next
      </Unknown>,
    ]
  }
  appendTo={<body />}
  ariaDescribedById=""
  className="ins-c-wizard"
  content={
    Array [
      <div
        id="first"
      >
        First
      </div>,
      <div
        id="second"
      >
        Second
      </div>,
      <div
        id="third"
      >
        Third
      </div>,
      <div
        id="fourth"
      >
        Fourth
      </div>,
    ]
  }
  hideTitle={false}
  isFooterLeftAligned={false}
  isLarge={true}
  isOpen={true}
  isSmall={false}
  onClose={[Function]}
  onKeyPress={[Function]}
  showClose={true}
  title="Some title"
>
  <div
    id="first"
  >
    First
  </div>
</Modal>
`;

exports[`Wizard component should render correctly one page 1`] = `
<Modal
  actions={
    Array [
      <Unknown
        action="cancel"
        onClick={[Function]}
        variant="secondary"
      >
        Cancel
      </Unknown>,
      false,
      <Unknown
        action="confirm"
        isDisabled={false}
        onClick={[Function]}
        variant="primary"
      >
        Confirm
      </Unknown>,
    ]
  }
  appendTo={<body />}
  ariaDescribedById=""
  className="ins-c-wizard"
  content={
    Array [
      <div>
        One
      </div>,
    ]
  }
  hideTitle={false}
  isFooterLeftAligned={false}
  isLarge={false}
  isOpen={true}
  isSmall={false}
  onClose={[Function]}
  onKeyPress={[Function]}
  showClose={true}
  title="Some title"
>
  <div>
    One
  </div>
</Modal>
`;
